style: Replace prettier with biome

rbm0524 opened the pull request:
### Description

#### Related issue
#5754 
We have applied the format of bio-format to js and ts. The function of bio-format is "Format the committed files". 
I tried to reflect '.pretiterignore' in the biome.json file, but
```
/code/src/js/biome.json configuration ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

× Found a nested root configuration, but there's already a root configuration.


i The other configuration was found in /code.


i Use the migration command from the root of the project to update the configuration.


$ biome migrate --write

configuration ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━



× Biome exited because the configuration resulted in errors. Please fix them.
```

Is there a way to get through pre-commit.ci even if I use the biome.json file?
I also want to consider whether it is necessary to migrate the contents of `.pretterignore` because the format is applied only to `js` and `ts`.

When I apply the formatter, I get an error in the CI, is there a formatting script or recommended settings that I use in this project?

 Please let me know if there is anything else we need to revise or discuss. 

#5818 branch is closed. I reflected what we discussed in that PR.

### Checklist

- [ ] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [ ] Add / update tests
- [ ] Add new / update outdated documentation

hoodmane commented:
This is very noisy. Is there a way to configure it so that the settings are more similar to prettier? I'd prefer to see a few hundred lines of stuff max. We probably should also keep the formatting in a separate commit to main so we can add it to .git-blame-ignore-revs .

ryanking13 commented:
> This is very noisy. Is there a way to configure it so that the settings are more similar to prettier? I'd prefer to see a few hundred lines of stuff max. We probably should also keep the formatting in a separate commit to main so we can add it to .git-blame-ignore-revs .

+1

It would be nice if we start with a similar formatting rule in this PR, and gradually applying more formating options

youknowone commented:
The most noisy part is switching from space to tab. Keeping space will make the diff reasonable size.

hoodmane commented:
`biome migrate prettier --write` apparently automatically migrates the config.

rbm0524 commented:
> `biome migrate prettier --write` apparently automatically migrates the config.

biome supports migration from `.prettierrc.json` to `biome.json.` It does not reflect '.prettierignore'.
There's no `prettierrc.json`. So, I think don't need to migration.

https://biomejs.dev/guides/big-projects/#monorepo
There was a recommended rule in this document, so I reflected it. Any other configurations do you want?

hoodmane commented:
We want to use the biome config that's as close as possible to the default prettier config. If it requires a `.prettierrc` file to do a migration just make an empty one.

rbm0524 commented:
> We want to use the biome config that's as close as possible to the default prettier config. If it requires a `.prettierrc` file to do a migration just make an empty one.

Migration problem is solved. 
```
× Found a nested root configuration, but there's already a root configuration.


i The other configuration was found in /code.
```

https://biomejs.dev/guides/big-projects/#monorepo
In this documents, 

>Create nested configuration files, one in each package where it is needed. These nested configuration files must have the field "root" set to false.
For convenience, when you use the microsyntax extends: "//", you can omit "root": false, because it is already implied that this configuration isn’t a root configuration:

I solved this problem through "root" set to false.

rbm0524 commented:
> It doesn't seem solved -- it is changing all the files to be indented with tabs instead of spaces. That needs to be fixed.

I created a `.prettierrc.json` and completed the migration. I think it's applied now, so please check.

ryanking13 commented:
Hi @rbm0524, thanks for your work.

It seems like we need to teach biome to use space instead of tab, I think we need `biome.json` starting with some basic configuration.

As you can see in the commit log, pre-commit.ci is keep auto fixing space to tabs.

<img width="950" height="49" alt="image" src="https://github.com/user-attachments/assets/1b96415f-485c-4615-bf44-025ac9b79df4" />

Here are some additional comments:

1. You don't need to add `.git-blame-ignore-revs` file in this PR, as we don't know the commit has until is is merged. That file can be added in a follow-up PR if this PR generates a large diff.
2. Let's remove `.prettierrc.json` and add `biome.json` instead. I think we can start with some basic configuration, as outlined in the [biome docs](https://biomejs.dev/guides/configure-biome/#_top), and then add more rules.
3. You can run `pre-commit run -a` locally to apply the formatting to the entire codebase regardless of commit. This will help you understand how formatting will be applied to the codebase.

rbm0524 commented:
> Hi @rbm0524, thanks for your work.
> 
> It seems like we need to teach biome to use space instead of tab, I think we need `biome.json` starting with some basic configuration.
> 
> As you can see in the commit log, pre-commit.ci is keep auto fixing space to tabs.
> 
> <img alt="image" width="950" height="49" src="https://private-user-images.githubusercontent.com/24893111/475446606-1b96415f-485c-4615-bf44-025ac9b79df4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTQ1NTQwNTEsIm5iZiI6MTc1NDU1Mzc1MSwicGF0aCI6Ii8yNDg5MzExMS80NzU0NDY2MDYtMWI5NjQxNWYtNDg1Yy00NjE1LWJmNDQtMDI1YWM5Yjc5ZGY0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODA3VDA4MDIzMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmZjY1YzZhMDQyODVmZWI0OGE3Nzc3MThmOGI2ZjA4MzU2MTFjY2Y0OTZkNDYxNDdiNTdmNjgxYjI4OTVmNWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2lqH6-XzsKi3m-Mfecop4pGNvgKFBWKRhkpT4bFTKBY">
> Here are some additional comments:
> 
> 1. You don't need to add `.git-blame-ignore-revs` file in this PR, as we don't know the commit has until is is merged. That file can be added in a follow-up PR if this PR generates a large diff.
> 2. Let's remove `.prettierrc.json` and add `biome.json` instead. I think we can start with some basic configuration, as outlined in the [biome docs](https://biomejs.dev/guides/configure-biome/#_top), and then add more rules.
> 3. You can run `pre-commit run -a` locally to apply the formatting to the entire codebase regardless of commit. This will help you understand how formatting will be applied to the codebase.

Removed `.prettierrc.json`. `biome.json` existed in the previous commit, but nested `biome.json` still seems to cause errors. So I moved `biome.json` to root.

hoodmane commented on biome.json:
```diff
@@ -0,0 +1,54 @@
+			"jsxQuoteStyle": "double",
+			"quoteProperties": "asNeeded",
+			"trailingCommas": "all",
+			"semicolons": "asNeeded",
```
I am strongly against semicolons being as needed.

rbm0524 replied:
I changed it to `always`.
Now, semicolons are always added at the end of each statement.

rbm0524 commented on biome.json:
```diff
@@ -0,0 +1,54 @@
+	"linter": { "enabled": false, "rules": { "recommended": true } },
+	"javascript": {
+		"formatter": {
+			"jsxQuoteStyle": "double",
```
I removed that

rbm0524 commented on biome.json:
```diff
@@ -0,0 +1,54 @@
+			"semicolons": "asNeeded",
+			"arrowParentheses": "always",
+			"bracketSameLine": false,
+			"quoteStyle": "single",
```
I changed it

rbm0524 commented:
> Looks great, thanks @rbm0524 . Now let's remove the ignore revs file and exclude pyodide-cross-build-environments.json and revert it and it'll be ready to merge.

I removed ignore revs file and revert `pyodide-cross-build-environments.json`
